hedera-node/docs/design/services/token-service/airdrops/token-claim-airdrop.md

```diff
@@ -0,0 +1,165 @@
+        - The transaction must be signed by the account referenced by a `receiver_id` for each entry in the pending airdrops list
+        - Confirm that for the given pending airdrops ids in the transaction there are corresponding pending transfers existing in state
+        - Check if the sender has sufficient balance to fulfill the airdrop
+    - Handle:
+        - Any additional validation depending on config or state i.e. semantics checks
+        - The business logic for claiming pending airdrops
+            - We need to create a token association between each `receiver_id` and `token_reference`, future rents for token association slot should be paid by `receiver_id`
+              - Since we would have the signature of the receiver, even if it's an account with `receiver_sig_required=true`, the claim would implicitly work properly
+            - Then we should transfer the claimed tokens to each `receiver_id`
+                - We can dispatch synthetic crypto transfer for this, but we must skip the assessment of custom fees
+                  - In case of a fungible token claim
+                    - Create a synthetic `CryptoTransfer` with the corresponding `sender`, `receiver`, `token` and `amount` based on the `PendingAirdropId` and the corresponding `PendingAirdropValue`
+                    - Then delegate it to the `CryptoTransfer.handler()` to execute the transfer
+                  - In case of an NFT claim
+                    - Create a synthetic `CryptoTransfer` with the corresponding `sender`, `receiver`, `token` and `serial number` based on the based on the `PendingAirdropId`
+                    - Then delegate it to `CryptoTransfer.handler()` to execute the transfer
+        - Token transfers and associations should be externalized using the `tokenTransferLists` and `automatic_token_associations` fields in the transaction record
+    - Fees calculation
+
+### Zero-Balance accounts
+
+An account with no open auto-association slots can receive airdrops but must send a `TokenClaimAirdrop` transaction, which requires a payer. If the account has zero hbars, then it can still claim the transfer if someone else is willing to pay for that transaction. For example, a Dapp could be the payer on the transaction. Both the Dapp and the account must sign the transaction.
+
+### Hollow accounts
+
+Any existing hollow accounts that were created before [HIP-904](https://hips.hedera.com/hip/hip-904) will have no or limited number of `maxAutoAssociations` depending on if they were created with HBAR or token transfer respectively. That means an airdrop of unassociated tokens to such accounts will result in a pending transfer.
+Performing `TokenClaimAirdrop` for such hollow account will also complete the account by setting its key which will obtained from the required transaction signature. Completing the hollow account should not modify the `maxAutoAssociations` on the account. That should always be an explicit step by a user.
```

Comment:
In general `used_auto_associations` should be incremented only if there are available auto-association slots i.e. `used_auto_associations < maxAutoAssociations`.
For those hollow accounts we should have `used_auto_associations == maxAutoAssociations`, so it would not make sense to increment `used_auto_associations`.

But completing the `TokenClaimAirdrop` will create an association with the claimed token and it will be externalized in `automatic_token_associations` field in transaction record, so that might be confusing.

Perhaps we should modify `maxAutoAssociations` and `automatic_token_associations` (they still should be equal, we just increment them) only for this specific case? @Nana-EC wdyt?

Reply:
Ah I see. I was thinking we should not modify `used_auto_associations`.